velox/functions/prestosql/ArrayCumSum.cpp

```diff
@@ -0,0 +1,175 @@
+    auto elementsRows =
+        toElementRows(elementsVector->size(), rows, arrayVector);
+    exec::LocalDecodedVector elements(context, *elementsVector, elementsRows);
+    auto rawElements = elementsVector->as<FlatVector<T>>()->mutableRawValues();
+    vector_size_t numRows = rows.end();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      auto size = arrayVector->sizeAt(row);
+      auto offset = arrayVector->offsetAt(row);
+      auto setNulls = [&](vector_size_t i) {
+        for (auto j = i; j < offset + size; j++) {
+          elementsVector->setNull(j, true);
+        }
+      };
+
+      if (elements->isNullAt(0)) {
+        setNulls(0);
+      } else {
+        for (auto i = offset + 1; i < offset + size; i++) {
+          if (elements->isNullAt(i)) {
+            setNulls(i);
+            break;
+          } else {
+            if constexpr (
+                std::is_same_v<T, Timestamp> || std::is_same_v<T, StringView> ||
+                std::is_same_v<T, bool>) {
+              VELOX_FAIL(
+                  "'+' cannot be applied to {}",
+                  CppToType<T>::create()->toString());
+            } else if constexpr (
+                std::is_same_v<T, float> || std::is_same_v<T, double>) {
+              rawElements[i] += rawElements[i - 1];
+            } else {
+              rawElements[i] =
+                  checkedPlus<T>(rawElements[i], rawElements[i - 1]);
+            }
+          }
+        }
+      }
+    });
+
+    return std::make_shared<ArrayVector>(
+        pool,
+        arrayVector->type(),
+        nullptr,
+        numRows,
+        arrayVector->offsets(),
+        arrayVector->sizes(),
+        arrayVector->elements());
+  }
+};
+
+void validateType(const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  VELOX_USER_CHECK_EQ(
+      inputArgs.size(), 1, "array_cum_sum requires exactly one parameter");
+
+  auto arrayType = inputArgs.front().type;
+  VELOX_USER_CHECK_EQ(
+      arrayType->kind(),
+      TypeKind::ARRAY,
+      "array_cum_sum requires arguments of type ARRAY");
+}
+
+template <TypeKind kind>
+std::shared_ptr<exec::VectorFunction> createTyped(
+    const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  VELOX_CHECK_EQ(inputArgs.size(), 1);
+
+  using T = typename TypeTraits<kind>::NativeType;
+  return std::make_shared<ArrayCumSumFunction<T>>();
+}
+
+std::shared_ptr<exec::VectorFunction> create(
+    const std::string& /* name */,
+    const std::vector<exec::VectorFunctionArg>& inputArgs,
+    const core::QueryConfig& /*config*/) {
+  validateType(inputArgs);
+  auto elementType = inputArgs.front().type->childAt(0);
+
+  return VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+      createTyped, elementType->kind(), inputArgs);
+}
+
+std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+  return {
+      // array(T) -> array(T)
+      exec::FunctionSignatureBuilder()
+          .typeVariable("T")
+          .returnType("array(T)")
```

Comment:
the signature should exclude unsupported types and only include supported ones.
you can create multiple signatures.

Reply:
Updated, thanks!